File: datastructures/queue/python/queue.py
from __future__ import annotations
from typing import Literal, Any
from dataclasses import dataclass


@dataclass
class Node:
    value: Any
    next: Node | None


@dataclass
class Queue:
    head: Node | None = None
    tail: Node | None = None
    length: Literal[0] = 0

    def enqueue(self, item: Any):
        """
        Add an element to the queue.

        O(1)
        """
        node = Node(value=item, next=None)
        if self.tail is None:
            self.tail = node
            self.head = node

        node.next = self.tail
        self.tail = node

    def peek(self) -> Any:
        if self.head:
            return self.head.value
        return self.head

File: datastructures/typescript/src/day1/Queue.ts
// npx jest .\Queue.ts

type Node<T> = {
    value:T,
    next?:Node<T>
}

export default class Queue<T> {
    public length: number;
    private head?:Node<T>;
    private tail?:Node<T>;
       

    constructor() {
        this.head = this.tail = undefined;
        this.length = 0;
    }
    
    // add item to the back of the queue:
    enqueue(item: T): void {
        const node = {value:item} as Node<T>;
        this.length++;
        if (!this.tail){
          this.tail = this.head = node;
          return
        }
        this.tail.next = node;
        this.tail = node;
    }
    // remove item from the queue
    deque(): T | undefined {
        if (!this.head){
            return undefined;
        }
        this.length--;
        const head = this.head;
        this.head = this.head.next;
        
        //free
        head.next = undefined;
        if (this.length ===0){
            this.tail = undefined;
        }
        return head.value

}
    peek(): T | undefined {
        return this.head?.value;

}
}